### PR TITLE
AAC mpeg4-generic audio

### PR DESCRIPTION
**Pull request by premultiply**

### CODE REVIEW

**Comment by premultiply** [comment]
Maybe the current (AAC-LATM) module should be renamed from `aac` to `aac-latm`.

**Comment by alfredh** [comment]
This is truly a quite impressive piece of work!

Thanks for working on this @premultiply 

I will do a more detailed review later, for now I only have
some comments:

1. some compiler warnings on OSX 10.14:

```
make aac.so
  CC [M]  build-x86_64/modules/aac/decode.o
In file included from modules/aac/decode.c:13:
modules/aac/aac.h:62:42: warning: no newline at end of file [-Wnewline-eof]
void aac_mirror_params(const char *fmtp);
                                         ^
modules/aac/decode.c:186:20: warning: implicit conversion loses integer precision:
      'unsigned long' to 'UINT' (aka 'unsigned int') [-Wshorten-64-to-32]
                bufferSize = len - pos;
                           ~ ~~~~^~~~~
2 warnings generated.
  CC [M]  build-x86_64/modules/aac/encode.o
In file included from modules/aac/encode.c:12:
modules/aac/aac.h:62:42: warning: no newline at end of file [-Wnewline-eof]
void aac_mirror_params(const char *fmtp);
                                         ^
modules/aac/encode.c:304:29: warning: implicit conversion loses integer precision:
      'unsigned long' to 'INT' (aka 'int') [-Wshorten-64-to-32]
                in_size = sizeof(int16_t) * (sampc - sampi);
                        ~ ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
modules/aac/encode.c:312:27: warning: implicit conversion loses integer precision:
      'unsigned long' to 'INT' (aka 'int') [-Wshorten-64-to-32]
                out_size = *len - total - AU_HDR_LEN; /* 1 au only! */
                         ~ ~~~~~~~~~~~~~^~~~~~~~~~~~
modules/aac/encode.c:320:32: warning: implicit conversion loses integer precision:
      'unsigned long' to 'INT' (aka 'int') [-Wshorten-64-to-32]
                in_args.numInSamples = sampc - sampi;
                                     ~ ~~~~~~^~~~~~~
4 warnings generated.
  CC [M]  build-x86_64/modules/aac/aac.o
In file included from modules/aac/aac.c:12:
modules/aac/aac.h:62:42: warning: no newline at end of file [-Wnewline-eof]
void aac_mirror_params(const char *fmtp);
                                         ^
1 warning generated.
  CC [M]  build-x86_64/modules/aac/sdp.o
In file included from modules/aac/sdp.c:12:
modules/aac/aac.h:62:42: warning: no newline at end of file [-Wnewline-eof]
void aac_mirror_params(const char *fmtp);
                                         ^
1 warning generated.
  LD [M]  aac.so
```

2. Testing baresip to baresip gives me only errors:

```
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 164 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 164 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 164 bytes: Protocol error
```

0x5 error is AAC_DEC_UNKNOWN

File: modules/aac/sdp.c [hunk 1]
```diff
@@ -26,30 +28,64 @@ static unsigned param_value(const char *fmtp, const char *name)
...
+
+	if (param_value(rfmtp, "indexLength") != INDEXLENGTH)
+		return false;
+
+	if (param_value(rfmtp, "indexDeltaLength") != INDEXDELTALENGTH)
+		return false;
+
+	if (param_value(rfmtp, "bitrate") < 8000 ||
+	    param_value(rfmtp, "bitrate") > 576000)
+		return false;
```
Comment (alfredh) [inline modules/aac/sdp.c]: is the `bitrate` parameter defined in RFC 3640 ?

if this parameter is missing, a sane default value should be used.

Comment (premultiply) [reply modules/aac/sdp.c]: Mmmmh...
I did not think about it at anytime as all of my test devices used it.

But if `bitrate` wont be there things may even get more complicated.

**Comment by premultiply** [comment]
Christmas vacations were coming to fast. ;-)
It was only compiled and tested with gcc on linux yet.
baresip2baresip has worked and should work without any problems.

**Comment by alfredh** [comment]
I tested with Linphone for iOS (Linphone version 4.2) which supports AAC.

The SDP from linphone to baresip looks like this:

```
v=0.
o=alfredh 2258 613 IN IP4 91.121.30.149.
s=Talk.
c=IN IP4 91.121.30.149.
t=0 0.
a=rtcp-xr:rcvr-rtt=all:10000 stat-summary=loss,dup,jitt,TTL voip-metrics.
m=audio 37720 RTP/AVP 96 101.
a=rtpmap:96 mpeg4-generic/48000.
a=fmtp:96 config=F8EE2000; constantDuration=512; indexDeltaLength=3; indexLength=3; mode=AAC-hbr; profile-level-id=76; sizeLength=13; streamType=5.
a=rtpmap:101 telephone-event/48000.
a=rtcp-fb:* trr-int 1000.
a=rtcp-fb:* ccm tmmbr.
a=nortpproxy:yes.
```

it does not really work, the audio is garbled and many decode errors:

```
answering call from sip:alfredh@sip.antisip.com with 200
stream: update 'audio'
audio: Set audio decoder: mpeg4-generic 48000Hz 1ch
aac: decode update: fmtp='config=F8EE2000; constantDuration=512; indexDeltaLength=3; indexLength=3; mode=AAC-hbr; profile-level-id=76; sizeLength=13; streamType=5'
audio: player started with sample format S16LE
audio: Set audio encoder: mpeg4-generic 48000Hz 1ch
aac: Encoder Profile AAC-ELD
rst: audio ptime=20 sampc=960 aubuf=[96000:1920000]
audio: source started with sample format S16LE
audio tx pipeline:         rst ---> mpeg4-generic
audio rx pipeline:   audiounit <--- mpeg4-generic
rst: connection established
rst: name='SLAY Radio' metaint=8192
stream: incoming rtp for 'audio' established, receiving from 91.121.30.149:37720
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 59 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 44 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 49 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 40 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 40 bytes: Protocol error
aac: aacDecoder_DecodeFrame() failed (0x5)
audio: mpeg4-generic codec decode 37 bytes: Protocol error
```

**Comment by premultiply** [comment]
Maybe your ptime is too long and the encoder has to package multiple AUs per packet?
This is NOT supported yet.

I did not find a way yet to request a specific audio packet length from baresips audio pipeline yet.
So this module needs ptime=1 at the moment.

File: modules/aac/sdp.c (hunk 1 above)
Comment (premultiply) [reply modules/aac/sdp.c]: Should be no problem to make the bitrate fmtp-property optional.
I think I remember that FDK-AAC will automatically select a proper bitrate if it is not specified explicit.

**Comment by alfredh** [comment]
I tested again with Linphone using AAC 22050Hz and it is working:

IMG_7546

here is output from baresip:

```
stream: incoming rtp for 'audio' established, receiving from 91.121.30.149:48558
.... aac decode: marker=0, len=92
.... header: offset = 4
             size = 88
             count = 1
.... aac decode: marker=0, len=51
.... header: offset = 4
             size = 47
             count = 1
.... aac decode: marker=0, len=49
.... header: offset = 4
             size = 45
             count = 1
.... aac decode: marker=0, len=43
.... header: offset = 4
             size = 39
             count = 1
rst: new format: 22050 hz, 1 ch, encoding 0x00d0
.... aac decode: marker=0, len=47
.... header: offset = 4
             size = 43
             count = 1
.... aac decode: marker=0, len=42
.... header: offset = 4
             size = 38
             count = 1
.... aac decode: marker=0, len=46
.... header: offset = 4
             size = 42
             count = 1
.... aac decode: marker=0, len=43
.... header: offset = 4
             size = 39
             count = 1
```

Marker bit from Linphone is always set to false.
Please note that I changed srate to 22050 and channels to 1.

Audio is working both ways, both encode and decode.

Regarding latm I would like to suggest that we register
two `struct aucodec`, one for mpeg4-generic and one for LATM.
then the same module can support for types.
I can update this code after the merge.

File: modules/aac/aac.c [hunk 2]
```diff
@@ -1,38 +1,299 @@
...
  *
  * Advanced Audio Coding (AAC) audio codec
+ *
+ * Supported version: libfdk-aac 0.1.6 or later
+ *
+ * Configuration options:
+ *
+ \verbatim
+  aac_samplerate     48000	# Encoded/decoded audio sample rate [Hz]
+  aac_samplerate         1	# Encoded/decoded audio channels
```
Comment (alfredh) [inline modules/aac/aac.c]: typo? should be aac_channels

File: modules/aac/aac.c [hunk 3]
```diff
@@ -1,38 +1,299 @@
...
-	.decupdh   = aac_decode_update,
-	.dech      = aac_decode_frm,
+	.name = "mpeg4-generic",
+	.encupdh = aac_encode_update,
+	.ench = aac_encode_frm,
+	.decupdh = aac_decode_update,
+	.dech = aac_decode_frm,
 	.fmtp_ench = aac_fmtp_enc,
 	.fmtp_cmph = aac_fmtp_cmp,
+	.ptime = 1, /* 480 samples per access unit @ 48000 hz */
```
Comment (alfredh) [inline modules/aac/aac.c]: please make the values aligned, just like the removed hunk (red)

File: modules/aac/aac.c [hunk 4]
```diff
@@ -1,38 +1,299 @@
...
+		}
+		break;
+	default:
+		warning("AAC Audio object types 2 (AAC-LC), 5 "
+			"(HE-AAC), 29 (HE-AAC v2), 23 (AAC-LD) "
+			"and 39 (AAC-ELD) are allowed.\n");
+		return EINVAL;
+	}
+
+	error = aacEncOpen(&enc, 0, 0);
```
Comment (alfredh) [inline modules/aac/aac.c]: I think the last parameter of aacEncOpen() takes "maxChannels"

File: modules/aac/aac.c [hunk 5]
```diff
@@ -1,38 +1,299 @@
...
+	error |= aacEncoder_SetParam(enc, AACENC_BITRATE, aac_bitrate);
+	error |= aacEncoder_SetParam(enc, AACENC_TRANSMUX, TT_MP4_RAW);
+	/* set object specific encoder params: */
+	error |= aacEncoder_SetParam(enc, AACENC_GRANULE_LENGTH,
+	                             aac_constantduration/aac_ratio);
+	error |= aacEncEncode(enc, NULL, NULL, NULL, NULL);
+	error |= aacEncInfo(enc, &enc_info);
+	if (error != AACENC_OK) {
+		return EINVAL;
+	}
```
Comment (alfredh) [inline modules/aac/aac.c]: this will leak memory. better to goto out and make sure "aacEncClose()" is called.

File: modules/aac/decode.c [hunk 6]
```diff
@@ -12,6 +13,13 @@
 #include "aac.h"
 
 
+struct au_hdr {
```
Comment (alfredh) [inline modules/aac/decode.c]: perhaps add a comment saying that "AU" means "Access Unit" ?
(au can also mean other things such as "audio")

File: modules/aac/encode.c [hunk 7]
```diff
@@ -30,117 +26,339 @@ static void destructor(void *arg)
...
-		goto out;
+	if (!aes) {
+		aes = mem_zalloc(sizeof(*aes), destructor);
+		if (!aes)
+			return ENOMEM;
 
-	aes = mem_zalloc(sizeof(*aes), destructor);
-	if (!aes)
-		return ENOMEM;
+		error = aacEncOpen(&aes->enc, 0, 0);
```
Comment (alfredh) [inline modules/aac/encode.c]: last parameter is "max channels".

Comment (premultiply) [reply modules/aac/encode.c]: Some encoding modes fail when parameter is not set to zero. Found it after hours of testing.

File: modules/aac/aac.c (hunk 2 above)
Comment (premultiply) [reply modules/aac/aac.c]: Right.

**Comment by alfredh** [comment]
Hi @premultiply I would like to merge the patch to master soon.

please note that we can also make changes to it, even after the merge ...

**Comment by premultiply** [comment]
I am back in the office next week and I am not able to test any changes before but i am fine with a merge now.
Bugs or other design issues may be addressed afterwards.

**Comment by premultiply** [comment]
Feel free to fix small issues left by yourself.

File: modules/aac/aac.c (hunk 5 above)
Comment (premultiply) [reply modules/aac/aac.c]: Are you sure?
There is no dynamic memory allocation in this case before.
This was taken from the example code at
https://github.com/mstorsjo/fdk-aac/blob/d47c2c625940d2bd62b0cd581605c5122c3d2cfd/aac-enc.c#L117

But I am not an expert in such questions.

**Pull request merged**